modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+    dir.create(folder.path)
+    # save corresponding block list to the folder.
+    blocks <- block.list[head.num:tail.num]
+    save(blocks, file = file.path(folder.path, "block.Rdata"))
+    # create job file.
+    jobsh <- readLines(con = system.file("analysis_qsub.job", package = "PEcAn.ModelName"), n = -1, warn=FALSE)
+    jobsh <- gsub("@FOLDER_PATH@", folder.path, jobsh)
+    writeLines(jobsh, con = file.path(folder.path, "job.sh"))
+    # qsub command.
+    qsub <- "qsub -l h_rt=48:00:00 -l buyin -pe omp 28 -V -N @NAME@ -o @STDOUT@ -e @STDERR@ -S /bin/bash"
+    qsub <- gsub("@NAME@", paste0("Job-", i), qsub)
+    qsub <- gsub("@STDOUT@", file.path(folder.path, "stdout.log"), qsub)
+    qsub <- gsub("@STDERR@", file.path(folder.path, "stderr.log"), qsub)
+    qsub <- strsplit(qsub, " (?=([^\"']*\"[^\"']*\")*[^\"']*$)", perl = TRUE)
+    cmd <- qsub[[1]]
+    out <- system2(cmd, file.path(folder.path, "job.sh"), stdout = TRUE, stderr = TRUE)
+    # grab job ids for future job completion detection.
+    job.ids <- c(job.ids, PEcAn.remote::qsub_get_jobid(
+      out = out[length(out)],
+      qsub.jobid = settings$host$qsub.jobid,
+      stop.on.error = TRUE))
+  }
+  # checking results.
+  PEcAn.logger::logger.info("Checking results.")
+  # if remaining number of jobs larger than 0.
+  while (length(job.ids) > 0) {
+    Sys.sleep(60)
+    completed_jobs <- job.ids %>% purrr::map(function(id) {
+      if (PEcAn.remote::qsub_run_finished(
+        run = id,
+        host = host,
+        qstat = qstat)) {
+        return(id)
+      }
+    }) %>% unlist()
+    job.ids <- job.ids[which(!job.ids %in% completed_jobs)]
+  }
+  # assemble results.
+  PEcAn.logger::logger.info("Assembling results.")
+  analysis <- c()
+  for (path in seq_along(folder.paths)) {
+    res_env <- new.env()
+    load(file.path(path, "results.Rdata"), envir = res_env)
+    analysis <- c(analysis, res_env$results)
+  }
+  return(analysis)
+}
+
+##' This function can help to execute `foreach` parallel MCMC sampling given generated MCMC configuration lists.
+##' @title qsub_analysis
+##' @param folder.path character: path where the `block.Rdata` file is stored.
+##' @param cores numeric: number of cpus used for parallel computaion. Default is NULL.
+##' @importFrom foreach %dopar%
+##' @export
+qsub_analysis <- function(folder.path, cores = NULL) {
+  # load file.
+  load(file.path(folder.path, "block.Rdata"))
+  # initialize parallel.
+  if (is.null(cores)) {
+    cores <- parallel::detectCores()
+  }
+  cl <- parallel::makeCluster(cores)
+  doSNOW::registerDoSNOW(cl)
+  # progress bar
+  pb <- utils::txtProgressBar(min=1, max=length(blocks), style=3)
+  progress <- function(n) utils::setTxtProgressBar(pb, n)
+  opts <- list(progress=progress)
+  # parallel computation.
+  l <- NULL # fix GitHub check issue.
+  results <- foreach::foreach(l = blocks, .packages=c("Kendall", "purrr"), .options.snow=opts) %dopar% {
+    MCMC_block_function(l)
+  }
```

Comment:
Naive question: How closely equivalent is this to `parallel::parLapply(cl, blocks, MCMC_block_function)`?  Would it be worth considering that approach since it only uses existing dependencies?

Reply:
So, I haven't used this parallel function before. The reason that I used `foreach` is that it has great management of memory usage with large data inputs. But I can try this function out to see if it performs similarly to the `foreach` package.